### ISSUE TITLE
Segfault when expanding 

### ISSUE THREAD

**Issue by lstrojny** [open-telemetry/opentelemetry-php-instrumentation#68]
PHP 8.2.7 with latest opentelemetry 1.0.0beta6.

**Steps to reproduce**
Given this code where we modify the arguments will crash with a segfault.

```php
<?php
OpenTelemetry\Instrumentation\hook(
    null,
    'array_slice',
    pre: function(null $instance, array $params) {
        return [$params[0], $params[1], 1];
    },
    post: function(null $instance, array $params, $returnValue) {
    }
);

var_dump(array_slice([1,2,3], 1));
```

This will not crash:

```diff
     null,
     'array_slice',
     pre: function(null $instance, array $params) {
-        return [$params[0], $params[1], 1];
+        return [$params[0], $params[1]];
     },
     post: function(null $instance, array $params, $returnValue) {
     }
```

This will also not crash:

```diff
     'array_slice',
     pre: function(null $instance, array $params) {
         return [$params[0], $params[1], 1];
-    },
-    post: function(null $instance, array $params, $returnValue) {
     }
 );
```

It will also not crash if the function is a userland function, so it only crashes if the function is defined in a PHP extension (core or PECL). I originally discovered the crash while working on instrumentation for [PECL amqp](https://pecl.php.net/amqp), specifically when hooking into `AMQPExchange::publish` and expanding the args there.

**Comment by brettmc** [comment]
I spent some time on this today. No fixes yet, but it seems to be related to garbage collection: the exception is in php shutdown. Whilst playing around, I also came across another failure scenario: we can't return $params from a pre callback.
The two may be related, as they're both to do with ref counting for garbage collection, I think. #71

**Comment by pdelewski** [comment]
@lstrojny @brettmc I'm wondering how common above scenario is, for me seems like corner case, however I might be wrong.

**Comment by brettmc** [comment]
I think it is a corner case. I believe it was a blocker for an instrumentation that Lars was playing with, but it doesn't seem to be an issue for any of our existing instrumentation modules. And, it's fairly obvious in testing/dev.

**Comment by lstrojny** [comment]
Sure, it's a corner case. That being said, I believe any extension, let alone an instrumentation extension, may never crash. 

It may not happen that a user installs a buggy instrumentation and suddenly a production workload starts crashing.

**Comment by pdelewski** [comment]
> Sure, it's a corner case. That being said, I believe any extension, let alone an instrumentation extension, may never crash.
> 
> It may not happen that a user installs a buggy instrumentation and suddenly a production workload starts crashing.

@lstrojny I totally agree with you. We are just trying to prioritize current set of bugs and work on them according to that.

**Comment by lstrojny** [comment]
Ignore my tirade then 😅

**Comment by brettmc** [comment]
"fixed" by #84 (ie, not fixed, but ignore+log so that it does not crash)

**Comment by brettmc** [comment]
I just re-checked this against `main` since a segfault bug was just fixed. Still broken.

**Comment by agoallikmaa** [comment]
I looked into this and it appears that at the end of each frame (after all arguments) is the pointer to the "previous observed frame". For internal functions, the pointer's location is calculated using the number of arguments. The value is written before the begin handler is invoked, and read after end handler is invoked.

Adding extra arguments both changes where it thinks the pointer is (as the argument count has changed), but more importantly, adding the argument itself overwrote the value at that address anyway. Looking at how it gets executed with `gdb` shows that it is reading it back later from a different address than it wrote it to:

```
Breakpoint 6, _zend_observe_fcall_begin (execute_data=0x7fef96c16190) at /php-src/Zend/zend_observer.c:235
235			*prev_observed_frame(execute_data) = current_observed_frame;
(gdb) print prev_observed_frame(execute_data)
$29 = (zend_execute_data **) 0x7fef96c16200

Breakpoint 7, zend_observer_fcall_end (execute_data=0x7fef96c16190, return_value=0x7fef96c160e0) at /php-src/Zend/zend_observer.c:284
284		current_observed_frame = *prev_observed_frame(execute_data);
(gdb) print prev_observed_frame(execute_data)
$30 = (zend_execute_data **) 0x7fef96c16210
```

The reason for the crash is that `current_observed_frame` is only read outside of these 2 locations during shutdown, to make sure end is called even if exiting within one of the observed functions without the function actually returning.

The reason why this does not seem to affect user functions at first is that there are some additional slots after arguments used for other purposes (called "temporary" and "compact" slots, minimum of 1 it seems), and the calculation to the pointer location uses the total number of "slots" with arguments and extra slots combined, which is held in a separate field that is not modified by the extension.

However, when adding many extra arguments to user functions, I was able to reproduce the crash for those too. This means that currently it is not safe to allow adding arguments for user functions either. Even if it doesn't crash, it means overwriting "slots" meant for other purposes, which means those may be read or written within the function itself for that other purpose, and a later attempt at reading the argument would read the new value rather than the argument value supplied by the begin handler.

The only safe solution I can think of is to calculate how many bytes of frame data is left after the argument part, and memcpy that entire region ahead based on how many arguments are added, and then update both the argument count and total slot count properly.

However, `zend_vm_stack_extend_call_frame` would have to be called then (user functions didn't crash so far because the argument was still written within the frame bounds). If the extended frame would not fit in the allocated stack, then `zend_vm_stack_extend_call_frame` allocates a new stack segment and copies the frame over to there. But the old frame location is still stored and used in Zend code via local variables of functions outside of our control. Then the only way to do it safely would be to somehow determine beforehand if the current stack segment could hold the extended frame, and only add the arguments if it can. This would be perceived as adding arguments "randomly failing" in the cases where the frame couldn't be expanded.

Based on this, it seems to me that the whole feature of adding arguments might not really be feasible. Unless the corner case where on a rare occasion the extra arguments just aren't added is acceptable.

**Comment by pdelewski** [comment]
@agoallikmaa Based on your investigation, whole feature seems to be tricky and agree with you that might not really be feasible. Maybe, we could emit some error message that would tell that such operation is not allowed.

**Comment by brettmc** [comment]
Thanks for the deep investigation! There is a todo left by the original author about maybe extending the call frame, but I trust your investigation that it will randomly fail.
Since we already do emit an error, I think the only other thing I'd like to do before closing this is link back to this issue from the source code and/or test...

**Comment by brettmc** [comment]
> The reason why this does not seem to affect user functions at first is that there are some additional slots after arguments used for other purposes (called "temporary" and "compact" slots, minimum of 1 it seems), and the calculation to the pointer location uses the total number of "slots" with arguments and extra slots combined, which is held in a separate field that is not modified by the extension.

So really, supporting expanding arguments at all is unsafe with our current implementation, and we should think about removing the functionality. At least then we're consistent.

**Comment by brettmc** [comment]
I think that [this test](https://github.com/open-telemetry/opentelemetry-php-instrumentation/blob/main/ext/tests/expand_params.phpt#L16) is bogus. It's not actually expanding parameters, it's just filling in a not-provided but expected parameter, which is not the same thing.
If I change the observed function to `function helloWorld($a)` then it does add an extra parameter, but it's got a NULL value - so, I think the functionality is broken anyway.

**Comment by lstrojny** [comment]
@agoallikmaa terrific investigation, well done. I can see two ways going forward:
 - Work with PHP core to find a way to reliable extend arguments (that might even require upstream changes and so it’s probably only going to land in 8.2 and 8.3)
 - Use an entirely different approach to instrumentation. One idea would be to implement `zend_ast_process()` and apply instrumentation by modifying the AST, another one could be to override `zend_execute` and apply instrumentation there

Anything I am missing?
